modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -89,86 +157,468 @@ public IgniteStatisticsRepository statisticsRepository() {
         if (F.isEmpty(colNames))
-            return columns;
+            return cols;
 
         Set<String> colNamesSet = new HashSet(Arrays.asList(colNames));
-        List<Column> resultList = new ArrayList<>(colNames.length);
+        List<Column> resList = new ArrayList<>(colNames.size());
 
-        for (Column col : columns)
+        for (Column col : cols)
             if (colNamesSet.contains(col.getName()))
-                resultList.add(col);
+                resList.add(col);
 
-        return resultList.toArray(new Column[resultList.size()]);
+        return resList.toArray(new Column[resList.size()]);
     }
 
     /** {@inheritDoc} */
-    @Override public void collectObjectStatistics(String schemaName, String objName, String... colNames)
-            throws IgniteCheckedException {
-        GridH2Table tbl = schemaMgr.dataTable(schemaName, objName);
-        if (tbl == null)
-            throw new IllegalArgumentException(String.format("Can't find table %s.%s", schemaName, objName));
+    @Override public void collectObjectStatistics(
+            String schemaName,
+            String objName,
+            String... colNames
+    ) throws IgniteCheckedException {
+        StatsCollectionFutureAdapter doneFut = new StatsCollectionFutureAdapter();
+        StatsKeyMessage keyMsg = new StatsKeyMessage(schemaName, objName, Arrays.asList(colNames));
+        UUID colId = UUID.randomUUID();
+        StatCollectionStatus status = new StatCollectionStatus(colId, Collections.singletonList(keyMsg),
+            Collections.emptyMap(), doneFut);
+        synchronized (status) {
+            currCollections.updateCollection(colId, s -> status);
+            Map<StatsKeyMessage, int[]> failedPartitions = null;
+            int cnt = 0;
+            do {
+                Collection<StatsCollectionAddrRequest> reqs = currCollections.generateCollectionRequests(colId,
+                    Collections.singletonList(keyMsg), failedPartitions);
+                Map<UUID, StatsCollectionRequest> msgs = reqs.stream().collect(Collectors.toMap(
+                    StatsCollectionAddrRequest::nodeId, StatsCollectionAddrRequest::req));
 
-        if (log.isDebugEnabled())
-            log.debug(String.format("Starting statistics collection by %s.%s object", schemaName, objName));
+                Map<UUID, StatsCollectionRequest> failedMsgs = sendLocalRequests(msgs);
+                failedPartitions = currCollections.extractFailed(failedMsgs.values()
+                    .toArray(new StatsCollectionRequest[0]));
+                if (cnt++ > 10)
+                    throw new IgniteCheckedException(String.format(
+                            "Unable to send all messages to collect statistics by key %s.%s", schemaName, objName));
+            }
+            while (!failedPartitions.isEmpty());
+        }
+        UUID locNode = ctx.discovery().localNode().id();
+        StatsCollectionAddrRequest locReq = status.remainingCollectionReqs().get(locNode);
```

Comment:
'status' doesn't reflect it's role here. 
It is actually a context of current task\operation.

Reply:
I would think we just created requests to 'reqs' map. but why we looking for local-node-request in status.remaining().
I can't figure out how we get there.

Reply:
Why 'status' doesn't fit here? Main purpouse of these function is just to send requests by specified status and update status with sended requests. So it's both status (in general it still represent the state of collection operation) and context (collectObjectStatistics update it with sended requests).
Your comment a little bit outdated in second part, but I'll describe current situation:
1) reqs  = currCollections.generateCollectionRequests(...   - we generate requests which should be sended ideally.
2) failedMsgs = sendRequests(reqs)    - we try to send as many requests as possible and return failed ones. BUT we didn't even try to send request to the local node. So it's no way to fail it.
3) sendedMsgs = reqs minus failed. - If local node should participate in collection - the request will be in sendedMsg
4) statsu.remainingCollectReqs().putAll(sendedMsgs) - will add local collection request with others to status, so we may get it from here.